test/test_type_promotion.py

```diff
@@ -0,0 +1,167 @@
+
+        expected = torch.ones([4, 4, 4], device=self.device)
+        expected = expected.to(torch.int32)
+
+        long_tensor = torch.ones([4, 4, 4], dtype=torch.int64, device=self.device)
+        int_tensor.add_(long_tensor)
+        int_tensor.add_(1)
+        three = expected + 2
+        self.assertEqual(int_tensor, three)
+        self.assertEqual(int_tensor.dtype, torch.int32)
+
+        byte_tensor = torch.tensor([1, 1, 1], dtype=torch.uint8)
+        half_tensor = torch.tensor([1, 1, 1], dtype=torch.half)
+        # byte is unsigned, half is signed.
+        self.assertRaisesRegex(RuntimeError, "can't be cast to", lambda: byte_tensor.add_(half_tensor))
+
+    # some basic examples
+
+    def test_int_promotion(self):
+        a = torch.ones([4, 4, 4], dtype=torch.int32, device=self.device)
+        b = torch.ones([4, 4, 4], dtype=torch.int64, device=self.device)
+        c = a + b
+        self.assertEqual(c, b + b)
+        self.assertEqual(c.dtype, torch.int64)
+
+    def test_float_promotion(self):
+        a = torch.ones([4, 4, 4], dtype=torch.float, device=self.device)
+        b = torch.ones([4, 4, 4], dtype=torch.double, device=self.device)
+        c = a + b
+        self.assertEqual(c, b + b)
+        self.assertEqual(c.dtype, torch.double)
+        c = b + a
+        self.assertEqual(c, b + b)
+        self.assertEqual(c.dtype, torch.double)
+
+    def test_add_wrapped(self):
+        a = torch.ones([4, 4, 4], dtype=torch.int, device=self.device)
+        b = 1
+        c = a + b
+        self.assertEqual(c, a + a)
+        self.assertEqual(c.dtype, torch.int)
+
+    def test_int_to_float(self):
+        a = torch.ones([4, 4, 4], dtype=torch.int32, device=self.device)
+        b = torch.ones([4, 4, 4], dtype=torch.float, device=self.device)
+        c = a + b
+        self.assertEqual(c.dtype, torch.float32)
+
+    # some examples from:
+    # https://github.com/pytorch/pytorch/issues/9515
+
+    def test_from_issue(self):
+        a = torch.rand(3, dtype=torch.float32, device=self.device)
+        u = torch.tensor([0, 0, 1], dtype=torch.uint8, device=self.device)
+        self.assertEqual((a * 5).dtype, torch.float32)
+        self.assertEqual((u + 1).dtype, torch.uint8)
+        self.assertEqual((u + 1000).dtype, torch.uint8)  # integer overflow
+
+        # not a "wrapped number"
+        other = torch.tensor(5.5, dtype=torch.double, device=self.device)
+
+        self.assertEqual((u + 5.5).dtype, torch.get_default_dtype())
+        self.assertEqual((u + other).dtype, torch.double)
+        # adding a 0-dim tensor to a float doesn't promote to double unless first
+        # type was integral.
+        self.assertEqual((a + other).dtype, torch.float32)
+
+    def test_half(self):
+        if(self.device == 'cpu'):
+            raise unittest.SkipTest('add_cpu not implemented for Half')
+        half = torch.tensor(5.5, dtype=torch.float16, device=self.device)
+        self.assertEqual((half + 2.2).dtype, torch.float16)
```

Comment:
This seems to follow what we also do for float, and avoids always converting to double.
We could potentially convert to float, but for mixed precision training this could give unexpected slowdowns I think